### PR TITLE
Allow message_attributes to be set from ActiveJob clients

### PR DESCRIPTION
**Pull request by timbreitkreutz**
We'd like to be able to access the SQS message attributes from the Shoryuken client code and middleware.  Previously, a memoized copy of the needed attributes were used which interfered with middleware when jobs were nested (i.e. ActiveJob clients calling other Active Jobs).  In addition, we had to work implement workarounds for metrics and tracing attributes at that level.

The suggestion here is to allow these to be passed in as options which then get used as the starting point.  The necessary active job attributes are then overwritten on top of any that are passed in to be sure they function the same as before.

We are open to any suggestions on style, thanks!

Co-authored by @tamouse

### CODE REVIEW

**Comment by timbreitkreutz** [comment]
@phstc Here's the nicer version of the PR we deleted a couple weeks ago. Is the utility function in code climate a blocker?  Is there anything else you'd like us to change or adjust on this one?

**Comment by timbreitkreutz** [comment]
Also how do we get Travis running?

**Comment by timbreitkreutz** [comment]
Update: We have been running this version in our main app for a couple weeks now and have no reported issues.

**Comment by phstc** [comment]
Hi @timbreitkreutz 

I can't really review this now. But I added a form for "recruiting" Shoryuken maintainers. Let me know if you are interested https://github.com/phstc/shoryuken/commit/4a03b64a4c48275d83c69b82c60c1379d9a350dc

File: lib/shoryuken/extensions/active_job_adapter.rb [hunk 1]
```diff
@@ -53,32 +53,25 @@ def calculate_delay(timestamp)
       def message(queue, job, options = {})
         body = job.serialize
 
-        msg = {}
+        attributes = options.delete(:message_attributes) || {}
+
+        msg = {
+          message_body: body,
+          message_attributes: attributes.merge(MESSAGE_ATTRIBUTES)
+        }
```
Comment (phstc) [inline lib/shoryuken/extensions/active_job_adapter.rb:61]: @timbreitkreutz Sorry for taking long to get back on this. 

Is it being used when enqueuing a job? `MyJob.perform_later(params)` 

I do miss the ability to send Shoryuken specific params (for example, FIFO params) when using Active Job is this change adding this support?

Comment (timbreitkreutz) [reply lib/shoryuken/extensions/active_job_adapter.rb:61]: Sorry was on vacation.  Will try to get an answer soon :)

Comment (timbreitkreutz) [reply lib/shoryuken/extensions/active_job_adapter.rb:61]: Sadly I don't think there is a way to inject this via `perform_later`--unless I'm missing something.  This PR was actually aimed more at the middleware layer.

Comment (cjlarose) [reply lib/shoryuken/extensions/active_job_adapter.rb:61]: @phstc I was able to test this locally. The proposed changes do not permit specification of `message_attributes` when using `perform_later` (I've documented that as a separate issue in #646).

Instead, the following can be used to enqueue a message with additional `message_attributes`, without overriding the `shoryuken_class` attribute.

```ruby
class AddJob < ActiveJob::Base
  def perform(a, b)
    puts a + b
  end
end

job = AddJob.new 1, 2
AddJob.queue_adapter.enqueue(job, message_attributes: { 'attr' => { string_value: 'myval', data_type: 'String' } })
```

Everything looks good from a testing perspective, so I'll merge it. Although it doesn't solve #646, it doesn't hinder it and might actually be an important prerequisite.

@timbreitkreutz Thanks so much for your patience on this one!

Comment (cjlarose) [reply lib/shoryuken/extensions/active_job_adapter.rb:61]: @timbreitkreutz If you could rebase on `master`, I'd appreciate it. Travis should run correctly after pushing up the rebased version. 🤞

Comment (timbreitkreutz) [reply lib/shoryuken/extensions/active_job_adapter.rb:61]: Thanks @cjlarose !

**Comment by timbreitkreutz** [comment]
Fork master has been updated.  Merge with squash I believe will result in a single commit to shoryuken repo?

**Comment by cjlarose** [comment]
Thanks! Tests are green. I'll squash merge.

**Comment by cjlarose** [comment]
Whoops. I said I'd squash merge and then immediately forgot. The `master` branch is protected from force-pushes, so I think I'll just leave it. An extra merge commit won't hurt anyone.

**Pull request merged**